src/irr/asset/CGraphicsPipelineLoaderMTL.cpp

```diff
@@ -288,28 +327,73 @@ void main()
         else
             Ns = PC.params.Ns;
 
-        color = Ka*Ia + Kd*NdotL;
+        vec3 diff = Kd*params.NdotL;
         switch (PC.params.extra&ILLUM_MODEL_MASK)
         {
+        case 1:
+            color = Ia*Ka + diff;
+            break;
         case 2:
-        case 3://2 + reflection map
-        case 4://3 with transparency (glass)
-        case 6:
-        case 8://reflection map
-        case 9://reflection map
-            color += Ks*pow(VdotR, Ns);
+        case 3://2 + IBL
+        case 5://basically same as 3
+        case 8://basically same as 5
+        {
+            float fr = Fresnel_dielectric(PC.params.Ni, params.NdotL);
+            vec3 spec = Ks*pow(params.NdotH, Ns);
+            color += Ia*Ka + mix(diff, spec, fr);
+        }
             break;
-        case 5:
+        case 4:
+        case 6:
         case 7:
-            color += Ks*pow(VdotR, Ns)*Fresnel_dielectric(PC.params.Ni, NdotL);
+        case 9://basically same as 4
+        {
+            float fr = Fresnel_dielectric(PC.params.Ni, params.NdotL);
+            vec3 spec = Ks*pow(params.NdotH, Ns);
+            color = fr*(Ia*Ka+spec);
+        }
             break;
         default:
             break;
         }
     }
     else color = Kd;
 
-    OutColor = vec4(color, d);
+    return color;  
+}
+
+#endif //_IRR_BSDF_COS_EVAL_DEFINED_
+
+
+
+#ifndef _IRR_FRAG_MAIN_DEFINED_
+#define _IRR_FRAG_MAIN_DEFINED_
+
+void main()
+{
+    irr_glsl_BSDFIsotropicParams bsdfParams = irr_glsl_computeLighting();
+    vec3 color = irr_glsl_bsdf_cos_eval(bsdfParams);
+
+    float d = PC.params.d;
+
+    //another illum model switch, required for illum=4,6,7,9 to compute alpha from fresnel (taken from opacity map or constant otherwise)
+    switch (PC.params.extra&ILLUM_MODEL_MASK)
+    {
+    case 4:
+    case 6:
+    case 7:
+    case 9:
+        d = Fresnel_dielectric(PC.params.Ni, bsdfParams.NdotL);
```

Comment:
(facepalm) the alpha (opacity) is `Fresnel(Ni, NdotV)`, why on earth would the fresnel transmittance depend on your light position? Its only the view position that determines how much you can see through.

P.S. This means that your blending equation needs to be a premultiplied alpha one (different fresnel factors on reflected color than just simply `1-transmission`), so pipeline needs blend state equivalent to:
```cpp
glBlendEquationSeparate(GL_FUNC_ADD, GL_FUNC_ADD);
glBlendFuncSeparate(GL_ONE, GL_ONE_MINUS_SRC_ALPHA, GL_ONE, GL_ONE_MINUS_SRC_ALPHA);
```

Reply:
idk, i wouldnt even think about it XD sorry